llvm/test/MC/Disassembler/X86/apx/movdir64b.txt

```diff
@@ -0,0 +1,10 @@
+# RUN: llvm-mc --disassemble %s -triple=x86_64 | FileCheck %s --check-prefixes=ATT
+# RUN: llvm-mc --disassemble %s -triple=x86_64 -x86-asm-syntax=intel --output-asm-variant=1 | FileCheck %s --check-prefixes=INTEL
+
+# ATT:   movdir64b	291(%r28d,%r29d,4), %r18d
+# INTEL: movdir64b	r18d, zmmword ptr [r28d + 4*r29d + 291]
+0x67,0x62,0x8c,0x79,0x08,0xf8,0x94,0xac,0x23,0x01,0x00,0x00
```

Comment:
I don't understand why `ATTR_ADSIZE` only matters to this instruction.
I think either all APX instructions need special handling for address size override prefix or nothing to do here.

I also compared with existing instruction that has `ATTR_EVEX` and `ATTR_OPSIZE`, e.g.,
```
EVEX.128.66.0F.W0 78 /r VCVTTPS2UQQ xmm1 {k1}{z}, xmm2/m64/m32bcst
```
We can decode correctly w/ and w/o address size override prefix without define a IC_EVEX_OPSIZE_ADSIZE, e.g.,
```
$ echo '0x62,0xf1,0x7d,0x0a,0x78,0x49,0x10' | llvm-mc --show-encoding --disassemble
        .text
        vcvttps2uqq     128(%rcx), %xmm1 {%k2}  # encoding: [0x62,0xf1,0x7d,0x0a,0x78,0x49,0x10]
$ echo '0x67,0x62,0xf1,0x7d,0x0a,0x78,0x49,0x10' | llvm-mc --show-encoding --disassemble
        .text
        vcvttps2uqq     128(%ecx), %xmm1 {%k2}  # encoding: [0x67,0x62,0xf1,0x7d,0x0a,0x78,0x49,0x10]
```

Does it still work with this patch?

Reply:
Oh, I see the reason why you considering it. We used `AdSize32` to distinguish the source register size of `movdir64b`, which is a special use of address size override prefix.
But I think the introduction of `i512mem_GR32` was just used to solve the problem. Does it not work in this case? I think we should follow this way and see what's missing for the new case.

Reply:
Ha! we have specail handling here https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/X86/Disassembler/X86Disassembler.cpp#L1326

Reply:
You're right. I introduce a new IC context named IC_EVEX_OPSIZE_ADSIZE for promoted movdir64b and just want disassembler to recognize the ADSIZE for movdir64b. But though IC_EVEX_OPSIZE_ADSIZE is just for movdir64b but the logic for ADSIZE in EVEX for disassembler could affect other instructions using ADSIZE. I may just need to expand the special handling for promoted movdir64b.
@phoebewang Thank you for pointing, will update.

Reply:
`IC_EVEX_OPSIZE_ADSIZE` is still needed as we added `IC_ADSIZE`, `IC_64BIT_REXW_ADSIZE`.  Dropping the change in X86Disassembler.cpp may solve the problem since we already specially handle it.

@XinWang10 Please add test mentioned by Phoebe in a separate commit, like https://reviews.llvm.org/D122449 

@phoebewang `i512mem_GR32` is added for decoding issue only? This surprised me...

Reply:
i512mem_GR32 could also for encoding, I added to make all the reg used by this instruction have same size.

Reply:
@KanRobert Test case will do. Thanks for info.

Reply:
@phoebewang This version could be more correct :)